### PR TITLE
[docs] fixed building at Python 2.x

### LINKED ISSUE #996: docs build fail on python 2
docs build fail on python 2
I repair it by change docs/conf.py line 30
from " from unittest.mock import Mock" to "from mock import Mock"
beceuse mock is a single model in python 2
i'm looking forward to someone can patch the tiny bug
thx.

## Environment info
Operating System: windows 10
CPU: intel core i7
C++/Python/R version:Python 2.7.13

## Error Message:

# Sphinx version: 1.6.4
# Python version: 2.7.13 (CPython)
# Docutils version: 0.14
# Jinja2 version: 2.9.6
# Last messages:

# Loaded extensions:
Traceback (most recent call last):
  File "c:\python27amd64\lib\site-packages\sphinx\cmdline.py", line 305, in main
    opts.warningiserror, opts.tags, opts.verbosity, opts.jobs)
  File "c:\python27amd64\lib\site-packages\sphinx\application.py", line 168, in __init__
    confoverrides or {}, self.tags)
  File "c:\python27amd64\lib\site-packages\sphinx\config.py", line 150, in __init__
    execfile_(filename, config)
  File "c:\python27amd64\lib\site-packages\sphinx\util\pycompat.py", line 150, in execfile_
    exec_(code, _globals)
  File "c:\python27amd64\lib\site-packages\six.py", line 709, in exec_
    exec("""exec _code_ in _globs_, _locs_""")
  File "<string>", line 1, in <module>
  File "conf.py", line 30, in <module>
    from unittest.mock import Mock
ImportError: No module named mock
## Reproducible examples
```
cd docs
./make html
```
## Steps to reproduce
0. install python 2
1.  cd docs
2.  ./make html

### PR DESCRIPTION
**Pull request by StrikerRUS**
fixed #996.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@henry0312 Do you have any ideas why it fails importing existing modules (only `sklearn` and `plotting` submodules) on OSX?
https://travis-ci.org/Microsoft/LightGBM/jobs/289277841#L1779

**Comment by henry0312** [comment]
I'm sorry, I don't know right now.

**Comment by StrikerRUS** [comment]
It fails even if I directly install `sklearn` and `matplotlib` https://github.com/StrikerRUS/LightGBM/commit/a287acce8941250383ce9b4d91d89602830f3da6.
OK, I'll try to find out the reason.

**Comment by StrikerRUS** [comment]
I've only discovered that it happens on Windows too. Only with Python 2.7.

**Comment by guolinke** [comment]
@StrikerRUS should we need to check docs in OSX ?

**Comment by StrikerRUS** [comment]
@guolinke Sorry, I was wrong - we have problems with Python 2.7 on all platforms. After finding the reason I'll reset Travis build matrix as it was before.

**Comment by StrikerRUS** [comment]
@guolinke May I delete `deprecation warning`?
https://github.com/Microsoft/LightGBM/blob/53b99854aa9046740e4ca12c401b28fb596e5d6d/python-package/lightgbm/sklearn.py#L585

It was introduced in #155 (Jan 3, version 0.1).

**Comment by guolinke** [comment]
@StrikerRUS 
I think @wxchan can answer you

**Comment by wxchan** [comment]
- -I am not sure. You can ask @ClimbsRocks 's opinion.

**Comment by StrikerRUS** [comment]
> Generally, I'd expect Deprecation warnings, or FutureWarnings, or UserWarnings for at least 2 minor versions (2.0.x - > 2.2.x), and ideally one major version, before a release that breaks existing codebases.

0.1 -> 2.0.11
I think that the condition is passed.

**Comment by guolinke** [comment]
@wxchan @henry0312 
please review this, thanks !

File: .travis/test.sh [hunk 1]
```diff
@@ -47,17 +53,15 @@ if [[ ${TASK} == "pylint" ]]; then
 fi
 
 if [[ ${TASK} == "if-else" ]]; then
-    conda create -q -n test-env python=$PYTHON_VERSION numpy
-    source activate test-env
+    pip install numpy
```
Comment (chivee) [inline .travis/test.sh]: using conda install should be fast ?

File: .travis/test.sh [hunk 2]
```diff
@@ -47,17 +53,15 @@ if [[ ${TASK} == "pylint" ]]; then
...
     mkdir build && cd build && cmake .. && make lightgbm || exit -1
     cd $TRAVIS_BUILD_DIR/tests/cpp_test && ../../lightgbm config=train.conf && ../../lightgbm config=predict.conf output_result=origin.pred || exit -1
     cd $TRAVIS_BUILD_DIR/build && make lightgbm || exit -1
     cd $TRAVIS_BUILD_DIR/tests/cpp_test && ../../lightgbm config=predict.conf output_result=ifelse.pred && python test.py || exit -1
     exit 0
 fi
 
-conda create -q -n test-env python=$PYTHON_VERSION numpy nose scipy scikit-learn pandas matplotlib pytest
-source activate test-env
+pip install numpy nose scipy scikit-learn pandas matplotlib pytest
```
Comment (chivee) [inline .travis/test.sh]: "Conda install"

File: docs/_static/js/rst_links_fix.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 $(function() {
     $('a[href^="./"][href*=".rst"]').attr('href', (i, val) => { return val.replace('.rst', '.html'); });  /* Replace '.rst' with '.html' in all internal links like './[Something].rst[#anchor]' */
     $('.wy-nav-content').each(function () { this.style.setProperty('max-width', 'none', 'important'); });
+    $('.wy-menu.wy-menu-vertical > ul:nth-of-type(2)').hide();  /* Fix theme navbar shows hidden toctree */
```
Comment (chivee) [inline docs/_static/js/rst_links_fix.js]: what is the second `<ul>` inlcuding? I only find the first `<ul>`with "Contents"

File: .travis/test.sh (hunk 1 above)
Comment (StrikerRUS) [reply .travis/test.sh]: I've made it for the uniformity because `rstcheck` isn't available through `conda install`.
Ok, I'll change the installation for all packages that are available through `conda install`.

File: docs/_static/js/rst_links_fix.js (hunk 3 above)
Comment (StrikerRUS) [reply docs/_static/js/rst_links_fix.js]: `sphinx_rtd_theme` renders a toctree which should be non-visible.
https://github.com/Microsoft/LightGBM/pull/997/files#diff-caf2a6b8f4947d018f68893c695b5202R27

http://www.sphinx-doc.org/en/stable/markup/toctree.html
>This will still notify Sphinx of the document hierarchy, but not insert links into the document at the location of the directive – this makes sense if you intend to insert these links yourself, in a different style, or in the HTML sidebar.

I don't know another way to avoid warnings about documents aren't included into any toctree.

Comment (StrikerRUS) [reply docs/_static/js/rst_links_fix.js]: From their documentation:
https://github.com/rtfd/sphinx_rtd_theme#how-the-table-of-contents-builds
>Also note that the table of contents is set with includehidden=true. This allows you to set a hidden toc in your index file with the hidden property that will allow you to build a toc without it rendering in your index.

**Comment by StrikerRUS** [comment]
@wxchan Please review the part with warnings.

**Comment by StrikerRUS** [comment]
@henry0312 Review please.

**Comment by henry0312** [comment]
I'll review soon.
I'm going to get about 2 weeks of vacation from tomorrow :)

**Comment by StrikerRUS** [comment]
Oh, it's great! 😃

File: python-package/lightgbm/compat.py [hunk 4]
```diff
@@ -63,7 +63,6 @@ class DataFrame(object):
     from sklearn.base import BaseEstimator
     from sklearn.base import RegressorMixin, ClassifierMixin
     from sklearn.preprocessing import LabelEncoder
-    from sklearn.utils import deprecated
```
Comment (henry0312) [inline python-package/lightgbm/compat.py]: why did you remove `deprecated`?

Comment (StrikerRUS) [reply python-package/lightgbm/compat.py]: It was the reason of failing on Python 2.
I replaced it with `warnings.warn('...', LGBMDeprecationWarning)` which is more universal mechanism and doesn't depend on sklearn.

Comment (henry0312) [reply python-package/lightgbm/compat.py]: I don't know why `deprecated` doesn't work with Python 2... even though scikit-learn supports Python 2.

Comment (StrikerRUS) [reply python-package/lightgbm/compat.py]: Yeah, it was strange for me too. I suppose it happens because `Mock()` didn't applied properly to this decorator in Python 2. Nevertheless, now we get rid of this headache.
